kube-controllers/pkg/controllers/node/ipam.go

```diff
@@ -568,6 +571,26 @@ func (c *ipamController) checkEmptyBlocks() error {
 			continue
 		}
 
+		// Get node to inspect labels
+		node, err := c.client.Nodes().Get(context.TODO(), nodeName, options.GetOptions{})
```

Comment:
We can't do a Node.Get() on every iteration here - for large clusters that's going to be a substantial increase in work performed by this controller and load on the API server.

 I believe we already have a watch configured that is caching node objects - we should be able to get the Node object from that cache instead.

Reply:
Here's where we created the IndexerInformer: https://github.com/projectcalico/calico/blob/3ca528bed2d9af244d70ca7cb581d0af107e148b/kube-controllers/pkg/controllers/node/controller.go#L54

You should be able pass it to the IPAM controller as a new argument, something like this:

```go
NewIPAMController(cfg, calicoClient, k8sClientset, nc.nodeInformer.Indexer())
```

Then, instead of this `Get()` command here, you can call the indexer's `GetByKey(<node name>)` command to return the cached node value.

Reply:
Thanks for the input. I changed it as you suggested.